filesystems/blob/src/main/scala/cromwell/filesystems/blob/BlobPathBuilderFactory.scala

```diff
@@ -1,50 +1,115 @@
-  val workspaceId: Option[String] = instanceConfig.as[Option[String]]("workspace-id")
-  val workspaceManagerURL: Option[String] = singletonConfig.config.as[Option[String]]("workspace-manager-url")
+  val container: BlobContainerName = BlobContainerName(instanceConfig.as[String]("container"))
+  val endpoint: EndpointURL = EndpointURL(instanceConfig.as[String]("endpoint"))
+  val workspaceId: Option[WorkspaceId] = instanceConfig.as[Option[String]]("workspace-id").map(WorkspaceId(_))
+  val expiryBufferMinutes: Long = instanceConfig.as[Option[Long]]("expiry-buffer-minutes").getOrElse(10)
+  val workspaceManagerURL: Option[WorkspaceManagerURL] = singletonConfig.config.as[Option[String]]("workspace-manager-url").map(WorkspaceManagerURL(_))
 
   val blobTokenGenerator: BlobTokenGenerator = BlobTokenGenerator.createBlobTokenGenerator(
     container, endpoint, workspaceId, workspaceManagerURL)
+  val fsm: BlobFileSystemManager = BlobFileSystemManager(container, endpoint, expiryBufferMinutes, blobTokenGenerator)
 
   override def withOptions(options: WorkflowOptions)(implicit as: ActorSystem, ec: ExecutionContext): Future[BlobPathBuilder] = {
     Future {
-      new BlobPathBuilder(blobTokenGenerator, container, endpoint)
+      new BlobPathBuilder(container, endpoint)(fsm)
     }
   }
 }
 
-sealed trait BlobTokenGenerator {
-  def getAccessToken: String
+final case class BlobContainerName(value: String) {override def toString: String = value}
+final case class StorageAccountName(value: String) {override def toString: String = value}
+final case class EndpointURL(value: String) {override def toString: String = value}
+final case class WorkspaceId(value: String) {override def toString: String = value}
+final case class WorkspaceManagerURL(value: String) {override def toString: String = value}
+
+object BlobFileSystemManager {
+  def parseTokenExpiry(token: AzureSasCredential): Option[Instant] = for {
+    expiryString <- token.getSignature.split("&").find(_.startsWith("se")).map(_.replaceFirst("se=","")).map(_.replace("%3A", ":"))
+    instant = Instant.parse(expiryString)
+  } yield instant
+
+  def buildConfigMap(credential: AzureSasCredential, container: BlobContainerName): Map[String, Object] = {
+    Map((AzureFileSystem.AZURE_STORAGE_SAS_TOKEN_CREDENTIAL, credential),
+      (AzureFileSystem.AZURE_STORAGE_FILE_STORES, container.value),
+      (AzureFileSystem.AZURE_STORAGE_SKIP_INITIAL_CONTAINER_CHECK, java.lang.Boolean.TRUE))
+  }
+  def hasTokenExpired(tokenExpiry: Instant, buffer: Duration): Boolean = Instant.now.plus(buffer).isAfter(tokenExpiry)
+  def uri(endpoint: EndpointURL) = new URI("azb://?endpoint=" + endpoint)
+}
+case class BlobFileSystemManager(container: BlobContainerName,
+    endpoint: EndpointURL,
+    expiryBufferMinutes: Long,
+    blobTokenGenerator: BlobTokenGenerator,
+    fileSystemAPI: FileSystemAPI = FileSystemAPI(),
+    initialExpiration: Option[Instant] = None) {
```

Comment:
What's the use case for passing in an initial expiry time? Feels kind of weird since there's no enforcement that the filesystem exists.

Reply:
This is mostly for testing to be honest, could make it a private arg possibly? I had been considering this or having a setter, but I liked a parameter that could only be set on construction better so that the internal timestamp wasn't something that could be easily modified.